### PR TITLE
Deprecate build listeners when CC is off

### PR DESCRIPTION
**Pull request by abstratt**
...regardless of the STABLE_CONFIGURATION_CACHE flag.

Issue: #22930

This changeset does the following:
- makes issuing of deprecation for build scope listeners not be dependent on STABLE_CONFIGURATION_CACHE
- introduces an internal option that allows disabling the issuing of deprecations for build scope listeners, to be leveraged by some tests
- changes tests that enabled STABLE_CONFIGURATION_CACHE so the build scope listener deprecations would be issued so they no longer did that, as that is no longer needed
- add expectation for deprecation warning for tests that use:
    -  `Gradle.buildFinished`
    - `Gradle.addListener`
    - `Gradle.addBuildListener`
    - `Gradle.useLogger`
    - `TaskExecutionGraph.beforeTask`
    - `TaskExecutionGraph.afterTask`
    - `TaskExecutionGraph.addTaskExecutionListener`
 - leverages the internal property introduced in a few console logging tests, as those get mixed with the expected output, and test cases don't have enough context to decide whether a deprecation should be expected or not.    
 - disables errors for deprecation warnings for some tests that hit the deprecated API
 - updates the `grgit` plugin to avoid a deprecation warning
 - add expected deprecations for Nebula smoke tests

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test ACC

**Comment by abstratt** [comment]
@bot-gradle test ACC

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test RFR

**Comment by abstratt** [comment]
@bot-gradle help

**Comment by abstratt** [comment]
@bot-gradle test RFR

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelBuildOperationsIntegrationTest.groovy [hunk 1]
```diff
@@ -163,12 +163,15 @@ This means workTask and slowTask would be expected to run concurrently in this c
...
-        workTaskDoesNotCompleteFirst()
+        workTaskDoesNotCompleteFirst(true)
     }
 
-    private void workTaskDoesNotCompleteFirst() {
+    private void workTaskDoesNotCompleteFirst(boolean expectDeprecation = false) {
         blockingHttpServer.expectConcurrent("workTask", "slowTask")
         args("--max-workers=4")
+        if (expectDeprecation) {
+            executer.expectDocumentedDeprecationWarning("Listener registration using Gradle.addListener() has been deprecated. This will fail with an error in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/current/userguide/upgrading_version_7.html#task_execution_events")
```
Comment (bamboo) [inline platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelBuildOperationsIntegrationTest.groovy:173]: 💅 It'd be clearer to extract this as method and call it from the places that pass `true` to the argument

File: platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt [hunk 2]
```diff
@@ -222,6 +229,11 @@ class KotlinDslPluginCrossVersionSmokeTest : AbstractKotlinIntegrationTest() {
         }
     }
 
+    private
+    fun expectBuildScopeListenerDeprecation() {
```
Comment (bamboo) [inline platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt:233]: ❓ Any chance of moving one of these utilities higher up and reuse wherever possible?

Comment (abstratt) [reply platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt:233]: We could, but that is not seems to be how we typically do it for tests. For tests, we seem to favor "code locality" over avoiding repetition across tests, at least for expected deprecations. I am usually bothered by repetition, but I can see the value of having the deprecation messages right there in the test case, or at least in the same file.

File: platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelBuildOperationsIntegrationTest.groovy (hunk 1 above)
Comment (abstratt) [reply platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelBuildOperationsIntegrationTest.groovy]: See: https://github.com/gradle/gradle/pull/30000/commits/93031caaa187f54a50768babfe37dcdd293e6faf

File: subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java [hunk 3]
```diff
@@ -121,12 +123,14 @@ public <T> T run(Function<? super BuildTreeLifecycleController, T> action) {
...
                 GradleInternal gradle = getBuildController().getGradle();
                 DefaultDeploymentRegistry deploymentRegistry = gradle.getServices().get(DefaultDeploymentRegistry.class);
-                gradle.addBuildListener(new InternalBuildAdapter() {
-                    @Override
-                    public void buildFinished(BuildResult result) {
-                        deploymentRegistry.buildFinished(result);
-                    }
-                });
+                DeprecationLogger.whileDisabled(() ->
+                    gradle.addBuildListener(new InternalBuildAdapter() {
```
Comment (bamboo) [inline subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java:127]: ❌ `InternalBuildAdapter` shouldn't trigger a deprecation warning since it's an `InternalBuildListener`

**Review by bamboo — DISMISSED** [review]

File: subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java (hunk 3 above)
Comment (abstratt) [reply subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java]: Please see https://github.com/gradle/gradle/pull/30000/commits/b38f0260a149fef41a79322168a2ea21bab3887d

File: subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildOperationsIntegrationTest.groovy [hunk 4]
```diff
@@ -418,6 +419,11 @@ class CompositeBuildOperationsIntegrationTest extends AbstractCompositeBuildInte
         buildB.buildFile << registration("buildB")
 
         when:
+        if (!(deprecation instanceof Wildcard)) {
```
Comment (bamboo) [inline subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildOperationsIntegrationTest.groovy:422]: TIL

File: platforms/documentation/docs/src/snippets/buildlifecycle/taskExecutionEvents/tests-kotlin/taskExecutionEvents.kotlin.sample.conf [hunk 5]
```diff
@@ -3,7 +3,8 @@
 # end::cli[]
 executable: gradle
 args: broken
-flags: --quiet
+# Do not fail for deprecation warnings: TaskExecutionGraph.beforeTask|afterTask
```
Comment (bamboo) [inline platforms/documentation/docs/src/snippets/buildlifecycle/taskExecutionEvents/tests-kotlin/taskExecutionEvents.kotlin.sample.conf:6]: ❓ Do we already have an issue about removing `beforeTask` and `afterTask`? We should have one and include a note about removing these samples as well.

File: platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt [hunk 6]
```diff
@@ -208,6 +212,9 @@ class KotlinDslPluginCrossVersionSmokeTest : AbstractKotlinIntegrationTest() {
                 "Consult the upgrading guide for further information: " +
                 "https://docs.gradle.org/current/userguide/upgrading_version_8.html#deprecated_access_to_conventions"
         )
+        if (!GradleContextualExecuter.isConfigCache()) {
```
Comment (bamboo) [inline platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt:215]: 💅 Favor `isNotConfigCache()` over `!isConfigCache()`
```suggestion
        if (GradleContextualExecuter.isNotConfigCache()) {
```

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy [hunk 7]
```diff
@@ -492,6 +492,9 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
                     "Consult the upgrading guide for further information: " +
                     "https://docs.gradle.org/current/userguide/upgrading_version_8.html#deprecated_access_to_conventions")
         }
+        if (kotlinVersionNumber < VersionNumber.parse("1.9.0") && !GradleContextualExecuter.isConfigCache()) {
```
Comment (bamboo) [inline platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:495]:
```suggestion
        if (kotlinVersionNumber < VersionNumber.parse("1.9.0") && GradleContextualExecuter.isNotConfigCache()) {
```

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/documentation/docs/src/snippets/buildlifecycle/taskExecutionEvents/tests-kotlin/taskExecutionEvents.kotlin.sample.conf (hunk 5 above)
Comment (abstratt) [reply platforms/documentation/docs/src/snippets/buildlifecycle/taskExecutionEvents/tests-kotlin/taskExecutionEvents.kotlin.sample.conf:6]: There is this: https://github.com/gradle/gradle/issues/22906

File: platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginGradlePluginCrossVersionSmokeTest.kt [hunk 8]
```diff
@@ -158,6 +158,9 @@ class KotlinDslPluginGradlePluginCrossVersionSmokeTest(
                     "https://docs.gradle.org/current/userguide/upgrading_version_8.html#deprecated_access_to_conventions"
             )
         }
+        if (kotlinVersion < VersionNumber.parse("1.9.0") && !GradleContextualExecuter.isConfigCache()) {
```
Comment (bamboo) [inline platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginGradlePluginCrossVersionSmokeTest.kt:161]:
```suggestion
        if (kotlinVersion < VersionNumber.parse("1.9.0") && GradleContextualExecuter.isNotConfigCache()) {
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r32/BuildFinishedCrossVersionSpec.groovy [hunk 9]
```diff
@@ -19,12 +19,17 @@ package org.gradle.integtests.tooling.r32
 import org.gradle.integtests.tooling.fixture.TargetGradleVersion
 import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
 import org.gradle.tooling.model.GradleProject
+import org.gradle.util.GradleVersion
 
 @TargetGradleVersion(">=3.2")
 class BuildFinishedCrossVersionSpec extends ToolingApiSpecification {
 
     def "model builders are executed before buildFinished hook"() {
+        println("releasedGradleVersion: " + targetVersion)
```
Comment (bamboo) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r32/BuildFinishedCrossVersionSpec.groovy:28]: 🤔 Is this change intentional or a leftover?

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy [hunk 10]
```diff
@@ -524,6 +527,9 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
...
```
Comment (bamboo) [inline platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:530]:
```suggestion
        if (kotlinVersionNumber < VersionNumber.parse("1.9.0") && GradleContextualExecuter.isNotConfigCache()) {
```

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r32/BuildFinishedCrossVersionSpec.groovy (hunk 9 above)
Comment (abstratt) [reply platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r32/BuildFinishedCrossVersionSpec.groovy:28]: oops, good catch

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test QF

**Comment by abstratt** [comment]
@bot-gradle test PRF

**Comment by abstratt** [comment]
@bot-gradle test PRF
